### PR TITLE
full & full_like: error on non-scalar fill_value

### LINKED ISSUE #6107: Expected behavior of dask.array.full() for non-scalar fill values?
What is the expected behavior for `dask.array.full()` for a non-scalar `fill_value`?

The `numpy.full()` signature in the documentation mentions "scalar" for `fill_value`.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.full.html

However, it'll broadcast non-scalar values if it can:
```python
a = np.full((3, 2), [1, 2])
```

Since ``dask.array.full()`` wraps this function, it works similarly:
```python
import dask.array

a1 = dask.array.full((2, 3), [[1, 2, 3], [4, 5, 6]])
a1.compute()
```

The issue is that when the chunk sizes change, it won't work anymore, as it starts broadcasting *per chunk*:
```python
a2 = dask.array.full((2, 3), [[1, 2, 3], [4, 5, 6]], chunks=(1, 3))
a2.compute()  # ValueError: could not broadcast input array from shape (2,3) into shape (1,3)
a3 = dask.array.full((2, 3), [1, 2, 3], chunks=(1, 3))
a3.compute()
```

My feeling is that choosing different chunk sizes shouldn't lead to different behavior here. I think I'd prefer a ValueError for a non-scalar fill_value.

### PR DESCRIPTION
**Pull request by Huite**
Fixes #6107
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

Because it touches the docstrings, I also ran `make html`, I didn't come across any issues.

I also found two nits to pick -- I think -- in the docstring:

First, the docstring template of `wrap` mentions:
>    Follows the signature of %(name)s exactly except that it also requires a  keyword argument chunks=(...)

The chunks argument is optional, not required. Maybe reword it to: 
> "... it also features an optional keyword argument chunks=(...)"

Second: technically another keyword argument is supported, `name`.

For this function:
https://github.com/dask/dask/blob/f58b95584dbdd19aad5c409892196f1e29a84534/dask/array/wrap.py#L14

`name` is explicitly extracted, if present, generated otherwise:
https://github.com/dask/dask/blob/f58b95584dbdd19aad5c409892196f1e29a84534/dask/array/wrap.py#L29-L33

So I think I'd have to add it to the function arguments.

Or is this form preferable?
```python
def full(shape, fill_value, *args, **kwargs):
```

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks. Noting that `chunks` and `name` are both optional arguments would be welcome. Do you want to do that it this pull reuqest?

File: dask/array/wrap.py [hunk 1]
```diff
@@ -146,4 +145,35 @@ def wrap(wrap_func, func, **kwargs):
...
+# Generate wrapped functions only once
+_full = w(np.full)
+_full_like = w_like(np.full, func_like=np.full_like)
+
+
+def full(shape, fill_value, dtype=None, order="C", chunks=None):
+    # np.isscalar has somewhat strange behavior:
+    # https://docs.scipy.org/doc/numpy/reference/generated/numpy.isscalar.html
+    if np.ndim(fill_value) != 0:
+        raise ValueError(f"fill_value must be scalar. Received {fill_value} instead.")
```
Comment (TomAugspurger) [inline dask/array/wrap.py:160]:
```suggestion
        raise ValueError(f"fill_value must be scalar. Received {type(fill_value)} instead.")
```

in case `fill_value` has a large string repr?

**Comment by Huite** [comment]
I just pushed another commit, it turns out the `name` argument wasn't supported by the the `*like` functions, the reason being that these are explicitly defined here:
https://github.com/dask/dask/blob/61eadd440736a5c46a0197450a5072646af916a2/dask/array/creation.py#L29

The wrapped forms of `*like` are not public, as far as I can tell:
```
grep -Hrn "from .wrap"
core.py:4627:    from .wrap import empty
creation.py:25:from .wrap import empty, ones, zeros, full
rechunk.py:24:from .wrap import empty
reductions.py:20:from .wrap import zeros, ones
routines.py:20:from .wrap import ones
__init__.py:220:    from .wrap import ones, zeros, empty, full
```

They are used internally: 
```
grep -Hrn "wrap\."
array/core.py:3490:            return wrap.empty_like(meta, shape=shape, chunks=shape, dtype=meta.dtype)
array/core.py:4277:            return wrap.empty_like(meta, shape=shape, chunks=shape, dtype=meta.dtype)
array/overlap.py:354:        c = wrap.full_like(
array/overlap.py:509:                empty = wrap.empty_like(
```

But appear to take a `meta` numpy array as their first argument.


Additionally, the `*like` functions in `creation` all follow this pattern:

```python
    a = asarray(a, name=False)
    return ones(
        a.shape,
        dtype=(dtype or a.dtype),
        chunks=(chunks if chunks is not None else a.chunks),
   )
```

It seems straightforward to include the `a = asarray(a, name=False)` line in the wrapping functions.

Currently, the `*like` functions in `creation` also do not support the memory `order` keyword, while the non-like functions do:

```python
In [8]: a = da.empty((10, 10), order="F")

In [9]: a.compute().flags
Out[9]:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

So I'm a little confused:
* Maybe rename the `wrap.*like` functions that are only used internally?
* Also make public the `*like` functions in `wrap`, and remove the repetition in `creation`? Due to `*args, **args` they would also support all numpy keywords like `order`.
* Alternatively, just make everything explicit in `creation` for all the wrapped functions, copy and edit the numpy docstrings. Taking the argument check of `full` into accoumt, 5/8 wrapped functions already have explicit definitions (`full, full_like, zeros_like, ones_like, empty_like`).

Adding `order` would be a breaking change for `*like` functions, as I think you'd want the numpy argument first, dask argument (`name`, `chunks`) last?

**Comment by TomAugspurger** [comment]
I think we'd be OK with adding an `order` argument to the public `da.*_like` methods.

> It seems straightforward to include the a = asarray(a, name=False) line in the wrapping functions.

That's probably fine, assuming the tests pass.

I'm less sure about your other suggestions / questions. There's a bit of repetition but it seems to be working ok.

**Comment by Huite** [comment]
Yeah, you're right. The latest commits are just the minimal changes necessary, as well as some minor polishes.

**Comment by jsignell** [comment]
@Huite is this good to go? Maybe with just one more look over from @TomAugspurger?

**Comment by jcrist** [comment]
Lets merge this, and I'll cleanup whatever else in #6276. Easier for me to fix merge conflicts than ask others to.
